File: cementerio-frontend/src/app/services/pago.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface PagoDTO {
  id?: number;
  monto: number;
  fecha?: string;
  concepto: string;
  estado: string; // PENDIENTE o PAGADO
  clienteId?: number;
  difuntoId?: number;
}

@Injectable({
  providedIn: 'root'
})
export class PagoService {
  private apiUrl = 'http://localhost:8081/api/pagos';

  constructor(private http: HttpClient) {}

  listarTodos(): Observable<PagoDTO[]> {
    return this.http.get<PagoDTO[]>(this.apiUrl);
  }

  listarPendientes(): Observable<PagoDTO[]> {
    return this.http.get<PagoDTO[]>(`${this.apiUrl}/pendientes`);
  }

  listarPorCliente(clienteId: number): Observable<PagoDTO[]> {
    return this.http.get<PagoDTO[]>(`${this.apiUrl}/cliente/${clienteId}`);
  }

  listarPorDifunto(difuntoId: number): Observable<PagoDTO[]> {
    return this.http.get<PagoDTO[]>(`${this.apiUrl}/difunto/${difuntoId}`);
  }

  registrarPago(pago: PagoDTO): Observable<PagoDTO> {
    return this.http.post<PagoDTO>(this.apiUrl, pago);
  }

  cambiarEstado(pagoId: number, estado: string): Observable<PagoDTO> {
    return this.http.put<PagoDTO>(`${this.apiUrl}/${pagoId}/estado?estado=${estado}`, {});
  }

  eliminarPago(pagoId: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${pagoId}`);
  }
}

File: Cementerio/src/main/java/com/cementerio/service/PagoService.java
package com.cementerio.service;

import com.cementerio.dto.PagoDTO;
import com.cementerio.entity.*;
import com.cementerio.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class PagoService {

    private final PagoRepository pagoRepository;
    private final ClienteRepository clienteRepository;
    private final DifuntoRepository difuntoRepository;

    private PagoDTO mapToDTO(Pago pago) {
        PagoDTO dto = new PagoDTO();
        dto.setId(pago.getId());
        dto.setMonto(pago.getMonto());
        dto.setFecha(pago.getFecha());
        dto.setConcepto(pago.getConcepto());
        dto.setEstado(pago.getEstado() != null ? pago.getEstado().name() : null);
        if (pago.getCliente() != null) dto.setClienteId(pago.getCliente().getId());
        if (pago.getDifunto() != null) dto.setDifuntoId(pago.getDifunto().getId());
        return dto;
    }

    public List<PagoDTO> listarTodos() {
        return pagoRepository.findAll().stream().map(this::mapToDTO).collect(Collectors.toList());
    }

    public List<PagoDTO> listarPorCliente(Long clienteId) {
        return pagoRepository.findByClienteId(clienteId).stream().map(this::mapToDTO).collect(Collectors.toList());
    }

    public List<PagoDTO> listarPorDifunto(Long difuntoId) {
        return pagoRepository.findByDifuntoId(difuntoId).stream().map(this::mapToDTO).collect(Collectors.toList());
    }

    public List<PagoDTO> listarPendientes() {
        return pagoRepository.findByEstado(EstadoPago.PENDIENTE).stream().map(this::mapToDTO).collect(Collectors.toList());
    }

    @Transactional
    public PagoDTO registrarPago(PagoDTO dto) {
        if (dto.getMonto() <= 0) {
            throw new RuntimeException("El monto debe ser mayor que cero");
        }

        Pago pago = new Pago();
        pago.setMonto(dto.getMonto());
        pago.setFecha(dto.getFecha() != null ? dto.getFecha() : LocalDate.now());
        pago.setConcepto(dto.getConcepto());
        pago.setEstado(dto.getEstado() != null && dto.getEstado().equalsIgnoreCase("PENDIENTE") ? EstadoPago.PENDIENTE : EstadoPago.PAGADO);

        if (dto.getClienteId() != null) {
            Cliente cliente = clienteRepository.findById(dto.getClienteId())
                    .orElseThrow(() -> new RuntimeException("Cliente no encontrado"));
            pago.setCliente(cliente);
        } else if (dto.getDifuntoId() != null) {
            Difunto difunto = difuntoRepository.findById(dto.getDifuntoId())
                    .orElseThrow(() -> new RuntimeException("Difunto no encontrado"));
            pago.setDifunto(difunto);
            // Si el pago está AL DÍA, reiniciamos el contador de fecha (renovación) del difunto público
            if (pago.getEstado() == EstadoPago.PAGADO) {
                // Renovar 7 años sumando a la fecha actual si ya estaba vencido, o sumando a la fecha de entierro?
                // En realidad solo se guarda el historial, la deuda desaparece si no hay pagos pendientes.
            }
        } else {
            throw new RuntimeException("Debe especificar un Cliente o un Difunto para el pago");
        }

        Pago saved = pagoRepository.save(pago);
        return mapToDTO(saved);
    }

    public PagoDTO cambiarEstado(Long pagoId, String nuevoEstado) {
        Pago pago = pagoRepository.findById(pagoId)
                .orElseThrow(() -> new RuntimeException("Pago no encontrado"));
        pago.setEstado(nuevoEstado.equalsIgnoreCase("PENDIENTE") ? EstadoPago.PENDIENTE : EstadoPago.PAGADO);
        return mapToDTO(pagoRepository.save(pago));
    }

    public void eliminar(Long pagoId) {
        if (!pagoRepository.existsById(pagoId)) {
            throw new RuntimeException("Pago no encontrado");
        }
        pagoRepository.deleteById(pagoId);
    }
}
